File: source/downloads/code/LeetCode/Python/Populating-next-right-pointers-in-each-node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        if root == None:
            return
        queue = []
        queue.append((root,0))
        oldnode = root
        oldlevel = 0
        while queue:
            item = queue.pop(0)
            cur = item[0]
            level = item[1]
            if cur.left != None:
                queue.append((cur.left,level+1))
            if cur.right != None:
                queue.append((cur.right,level+1))
            
            if oldnode != cur and oldlevel == level:
                oldnode.next = cur
            oldlevel = level
            oldnode = cur
        return

File: source/downloads/code/LeetCode/Java/Populating-Next-Right-Pointers-in-Each-Node.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        if(root==null)
            return ;
        
        TreeLinkNode cur = null;
        TreeLinkNode prev = root;
        while(prev.left != null){
            cur = prev;
            // 没一层都连起来
            while(cur!=null){
                cur.left.next = cur.right;
                if(cur.next!=null)
                    cur.right.next = cur.next.left;
                cur = cur.next;
            }

            // 下一层
            prev = prev.left;
        
        }   
    
    }
}
